### PR TITLE
rpc: increased log level in case RPC request could not be read

### PR DESCRIPTION
**Pull request by bas-vk**
This PR increases the log level in the RPC server when the next RPC request could not be read. This helps with trouble shooting (primarily mist related) connectivity issues.

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Fri Apr 15 09:49:17 UTC 2016_

File: rpc/errors.go [hunk 1]
```diff
@@ -95,3 +100,15 @@ func (e *shutdownError) Code() int {
 func (e *shutdownError) Error() string {
 	return "server is shutting down"
 }
+
+// connectionClosedError is thrown when the connection is closed
+type connectionClosedError struct {
+}
+
+func (e *connectionClosedError) Code() int {
```
Comment (karalabe) [inline rpc/errors.go]: Where is this code used? If it's needed to satisfy some interface somewhere, it would be nice to add it as a doc.

File: rpc/errors.go [hunk 2]
```diff
@@ -18,6 +18,11 @@ package rpc
 
 import "fmt"
 
+var (
+	// ErrConnectionClosed is thrown when the connection is closed
+	ErrConnectionClosed = new(connectionClosedError)
```
Comment (karalabe) [inline rpc/errors.go]: Is an entire new type really needed? Can't we just do `errors.New` here? If no, I'd rather make the type a tad more generic because currently the type itself is constant too (i.e. code and error message not configurable).

File: rpc/errors.go (hunk 1 above)
Comment (fjl) [inline rpc/errors.go]: It implements the `RPCError` interface in the same package. IMHO it is not necessary to document that it implements this interface.

File: rpc/errors.go (hunk 2 above)
Comment (fjl) [inline rpc/errors.go]: Maybe it would be nice to have `func newError(code int, format string, args ...interface{}) RPCError`. It could be used to remove more error types which only exist to satisfy the interface.

File: rpc/json.go [hunk 3]
```diff
@@ -119,6 +119,9 @@ func (c *jsonCodec) ReadRequestHeaders() ([]rpcRequest, bool, RPCError) {
 
 	var incomingMsg json.RawMessage
 	if err := c.d.Decode(&incomingMsg); err != nil {
+		if err == io.EOF {
+			return nil, false, ErrConnectionClosed
```
Comment (fjl) [inline rpc/json.go]: Why not return `io.EOF`?

File: rpc/errors.go (hunk 1 above)
Comment (bas-vk) [inline rpc/errors.go]: I actually begin renaming this to `rpc.Error` since `rpc.RPCError` interface is a bit ambiguous. But in the process I decided to make a separate PR for it because there are more things that can be renamed and I didn't want to clutter this PR.

Not sure about renaming this method to `ErrorCode`. It would be something like `rpc.RPCError.ErrorCode`.

Comment (fjl) [inline rpc/errors.go]: > It would be something like rpc.RPCError.ErrorCode.

That's a weird way of thinking about it. The only place where the method name is
visible is when defining a method.

Compare

```go
func (someType) Code() int { ... }
```

with 

```go
func (someType) ErrorCode() int { ... }
```.

Which one tells you more about the purpose of this method?

File: rpc/errors.go (hunk 2 above)
Comment (bas-vk) [inline rpc/errors.go]: The same error can be thrown in multiple places. For example parsing arguments can throw the same `invalidParamsError` errors when:

1. at least one of the arguments has an invalid type
2. not enough arguments
3. too many arguments

If we introduce a special helper function we must be aware that in all cases the same error code is returned. This is the reason for the special types, to get consistency.

File: rpc/json.go (hunk 3 above)
Comment (bas-vk) [inline rpc/json.go]: The codec returns an `RPCError` since this method can return multiple type of errors with different RPC error codes.

**Comment by fjl** [comment]
:+1: we can discuss the stuff that was brought up here later.

**Comment by fjl** [comment]
@bas-vk do you still want to get this merged?

**Comment by bas-vk** [comment]
I will close this one and create a new PR that will cleanup the RPC package.
For Mist we have a new connectivity layer that allows for better tracing.
